{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from project import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import csv\n",
    "from shared import *\n",
    "from time import time\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_reads():\n",
    "    \"\"\"\n",
    "    Reads all the RNA reads from reads.fa and returns a list of sequences as strings.\n",
    "\n",
    "    :return: a list of RNA sequence reads as strings\n",
    "    \"\"\"\n",
    "    max_id_len = 600\n",
    "    reads = np.zeros((1575, 2), dtype='U{:d}'.format(max_id_len))\n",
    "    i = 0\n",
    "    for seq_record in SeqIO.parse(\"reads.fa\", \"fasta\"):\n",
    "        reads[i, 0] = seq_record.id\n",
    "        reads[i, 1] = str(seq_record.seq)\n",
    "        i += 1\n",
    "    return reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_genome():\n",
    "    \"\"\"\n",
    "    Reads the genome sequence from genome.fa and return the genome sequence as a string.\n",
    "\n",
    "    :return: a string of the genome sequence\n",
    "    \"\"\"\n",
    "    genome = None\n",
    "    for seq_record in SeqIO.parse(\"genome.fa\", \"fasta\"):\n",
    "        genome = str(seq_record.seq)\n",
    "    return genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_known_genes():\n",
    "    \"\"\"\n",
    "    Reads the un/known genes, isoforms, and exons from genes.tab and constructs objects for each\n",
    "    and return the list of constructed genes.\n",
    "\n",
    "    :return: known_genes (a list of known Gene objects), unknown_genes (a list of unknown Gene objects)\n",
    "    \"\"\"\n",
    "    known_genes, known_isoforms, known_exons = {}, {}, {}\n",
    "    unknown_genes, unknown_isoforms, unknown_exons = {}, {}, {}\n",
    "    with open(\"genes.tab\") as tsv:\n",
    "        for line in csv.reader(tsv, dialect=\"excel-tab\"):\n",
    "            name = line[0]\n",
    "            if name == 'gene':\n",
    "                known_genes[line[1]] = line[2].split(';')\n",
    "            elif name == 'isoform':\n",
    "                known_isoforms[line[1]] = line[2].split(';')\n",
    "            elif name == 'exon':\n",
    "                known_exons[line[1]] = Exon(line[1], int(line[2]), int(line[3]))\n",
    "            elif name == 'unknown_gene':\n",
    "                unknown_genes[line[1]] = line[2].split(';')\n",
    "            elif name == 'unknown_isoform':\n",
    "                unknown_isoforms[line[1]] = line[2].split(';')\n",
    "            elif name == 'unknown_exon':\n",
    "                unknown_exons[line[1]] = Exon(line[1], int(line[2]), int(line[3]))\n",
    "    # Create the known Isoform objects\n",
    "    for k in known_isoforms:\n",
    "        known_isoforms[k] = Isoform(k, [known_exons[key] for key in known_isoforms[k]])\n",
    "        \n",
    "    # Create the UNknown Isoform objects\n",
    "    for k in unknown_isoforms:\n",
    "        unknown_isoforms[k] = Isoform(k, [unknown_exons[key] for key in unknown_isoforms[k]])\n",
    "\n",
    "    # Create the known Genes objects\n",
    "    for k in known_genes:\n",
    "        known_genes[k] = Gene(k, [known_isoforms[key] for key in known_genes[k]])\n",
    "        \n",
    "    # Create the UNknown Genes objects\n",
    "    for k in unknown_genes:\n",
    "        unknown_genes[k] = Gene(k, [unknown_isoforms[key] for key in unknown_genes[k]])\n",
    "    return known_genes, unknown_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main\n",
    "reads = read_reads()\n",
    "genome_sequence = read_genome()\n",
    "known_genes, unknown_genes = read_known_genes()\n",
    "known_genes_list = []\n",
    "for gene in known_genes:\n",
    "    known_genes_list.append(known_genes[gene])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "392.3921148777008"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = -time()\n",
    "aligner = Aligner(genome_sequence, known_genes_list)\n",
    "time() + t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evaluation import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "known_isoforms = []\n",
    "for gene in known_genes:\n",
    "    known_isoforms.extend(known_genes[gene].isoforms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "unknown_isoforms = []\n",
    "for gene in unknown_genes:\n",
    "    unknown_isoforms.extend(unknown_genes[gene].isoforms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_isoform_offsets = index_isoform_locations(known_isoforms, unknown_isoforms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0, 6759550, 50)]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_sequence = reads[150][1]\n",
    "alignment = aligner.align(reads[150][1])\n",
    "alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       0\n",
      "       1\n",
      "       2\n",
      "       2\n",
      "       unaligned alignment\n",
      "       [] GTGACTTAGCTGCAAAACCTACTGCGAGTGATGCCGTGGCTCCGATAGGT\n",
      "       3\n",
      "       3\n",
      "       unaligned alignment\n",
      "       [] TCTCGGGGTGAATACCTCTTATCGCGATACCTCCGGGGACTAGTGCGCCA\n",
      "       4\n",
      "       5\n",
      "       6\n",
      "       7\n",
      "       8\n",
      "       8\n",
      "       unaligned alignment\n",
      "       [] TAATCAGAAGGTGGATCAACTGGAAGATGTGCCTCCTCCAAAGAGCCGTA\n",
      "       9\n",
      "       10\n",
      "       11\n",
      "       12\n",
      "       13\n",
      "       14\n",
      "       15\n",
      "       16\n",
      "       17\n",
      "       18\n",
      "       19\n",
      "       20\n",
      "       21\n",
      "       22\n",
      "       23\n",
      "       24\n",
      "       25\n",
      "       26\n",
      "       27\n",
      "       28\n",
      "       29\n",
      "       29\n",
      "       unaligned alignment\n",
      "       [] AGCTTTGAAGAGACTACTGTCGATTTCTCCTGTCGCTGTCCGTGCGGACT\n",
      "       30\n",
      "       31\n",
      "       32\n",
      "       33\n",
      "       34\n",
      "       35\n",
      "       36\n",
      "       37\n",
      "       38\n",
      "       39\n",
      "       40\n",
      "       41\n",
      "       42\n",
      "       43\n",
      "       44\n",
      "       45\n",
      "       46\n",
      "       47\n",
      "       48\n",
      "       49\n",
      "       50\n",
      "       51\n",
      "       52\n",
      "       53\n",
      "       54\n",
      "       55\n",
      "       56\n",
      "       57\n",
      "       58\n",
      "       59\n",
      "       60\n",
      "       61\n",
      "       61\n",
      "       unaligned alignment\n",
      "       [] GTAAGTTGATATTTGGAGCCAATCTACTTAAGAACAAAATTGCCATTGCA\n",
      "       62\n",
      "       63\n",
      "       64\n",
      "       65\n",
      "       66\n",
      "       67\n",
      "       68\n",
      "       69\n",
      "       70\n",
      "       71\n",
      "       72\n",
      "       73\n",
      "       74\n",
      "       75\n",
      "       76\n",
      "       77\n",
      "       78\n",
      "       79\n",
      "       80\n",
      "       81\n",
      "       82\n",
      "       83\n",
      "       84\n",
      "       84\n",
      "       unaligned alignment\n",
      "       [] ATAAGAACATTCATCTGGTAGTTCTTGGAAGATGTCCCACCTCGAAAGAG\n",
      "       85\n",
      "       86\n",
      "       87\n",
      "       88\n",
      "       89\n",
      "       90\n",
      "       91\n",
      "       91\n",
      "       hidden_gene alignment\n",
      "       [(0, 2590477, 50)] AGGAAATTGAAAGTGTGAAAGAAAACACTGATAAACTTCTAACGGCTATC\n",
      "       92\n",
      "       93\n",
      "       94\n",
      "       95\n",
      "       96\n",
      "       97\n",
      "       98\n",
      "       99\n",
      "       100\n",
      "       101\n",
      "       102\n",
      "       103\n",
      "       104\n",
      "       104\n",
      "       hidden_gene alignment\n",
      "       [(0, 2559519, 50)] ATGGGCAAGTGGTGCTGTCACTGCTTCCACTGCTGCAGGGGTAGCGGCAA\n",
      "       105\n",
      "       106\n",
      "       107\n",
      "       108\n",
      "       109\n",
      "       110\n",
      "       111\n",
      "       112\n",
      "       113\n",
      "       114\n",
      "       115\n",
      "       116\n",
      "       117\n",
      "       118\n",
      "       119\n",
      "       120\n",
      "       121\n",
      "       122\n",
      "       123\n",
      "       124\n",
      "       125\n",
      "       126\n",
      "       127\n",
      "       128\n",
      "       129\n",
      "       130\n",
      "       131\n",
      "       132\n",
      "       133\n",
      "       134\n",
      "       135\n",
      "       135\n",
      "       unaligned alignment\n",
      "       [] GCCGAACGCAATTACCTCACGCCCTGCCTCATTCGTTGGCACGGGGAACC\n",
      "       136\n",
      "       137\n",
      "       138\n",
      "       139\n",
      "       140\n",
      "       141\n",
      "       142\n",
      "       143\n",
      "       144\n",
      "       145\n",
      "       146\n",
      "       147\n",
      "       148\n",
      "       149\n",
      "       150\n",
      "       151\n",
      "       152\n",
      "       153\n",
      "       154\n",
      "       154\n",
      "       unaligned alignment\n",
      "       [] GAATTCCCAGTGTCCAGAGCTGTGAAAAAAGATGTTCAATAGGAAAAAGG\n",
      "       155\n",
      "       156\n",
      "       157\n",
      "       158\n",
      "       159\n",
      "       160\n",
      "       161\n",
      "       162\n",
      "       163\n",
      "       163\n",
      "       unaligned alignment\n",
      "       [] CCGTGAGCCCTTACAGAGCCACGATGAAAATGTTGAGGAGAACTCAACTT\n",
      "       164\n",
      "       165\n",
      "       166\n",
      "       167\n",
      "       167\n",
      "       hidden_gene alignment\n",
      "       [(0, 6455646, 50)] GTCGCCCAGGCTGGAGTGCAGTAGCGAGATCCCGGCTCCCGGCAACCTCT\n",
      "       168\n",
      "       169\n",
      "       170\n",
      "       171\n",
      "       172\n",
      "       173\n",
      "       173\n",
      "       hidden_gene alignment\n",
      "       [(0, 3920513, 50)] ATTATGTTCATTTTTAAAAAGGGATTTGATAGACAAGATCTAGAGAAACA\n",
      "       174\n",
      "       175\n",
      "       176\n",
      "       177\n",
      "       178\n",
      "       179\n",
      "       179\n",
      "       unaligned alignment\n",
      "       [] GCGATATGGAGGCTAACTCTCAAACGTTTCTGATCATGCGGTCCTATTCG\n",
      "       180\n",
      "       181\n",
      "       182\n",
      "       183\n",
      "       184\n",
      "       185\n",
      "       186\n",
      "       187\n",
      "       188\n",
      "       189\n",
      "       190\n",
      "       191\n",
      "       192\n",
      "       193\n",
      "       194\n",
      "       195\n",
      "       196\n",
      "       197\n",
      "       198\n",
      "       199\n",
      "       200\n",
      "       200\n",
      "       unaligned alignment\n",
      "       [] TGCGATCGTTGCAGCCTTGCCATAGGATAAGAATTAGAGCGCAGATGCTC\n",
      "       201\n",
      "       201\n",
      "       unaligned alignment\n",
      "       [] GATGGAATAATATGAACACCAAACCCTTGTCTTCATATGGATTACTGGGG\n",
      "       202\n",
      "       203\n",
      "       204\n",
      "       205\n",
      "       206\n",
      "       207\n",
      "       208\n",
      "       209\n",
      "       210\n",
      "       211\n",
      "       212\n",
      "       213\n",
      "       213\n",
      "       unaligned alignment\n",
      "       [] CTGCTCTGCTTCGTGCTCCTGTGCGGAGTAGTGGTTTCGCCAGAAGTTAG\n",
      "       214\n",
      "       215\n",
      "       216\n",
      "       217\n",
      "       218\n",
      "       218\n",
      "       unaligned alignment\n",
      "       [] CTGATAAGCAGCAGGTGGATCAAGTGTTTTTTTAGATTCTGGAGACAAAA\n",
      "       219\n",
      "       220\n",
      "       221\n",
      "       222\n",
      "       223\n",
      "       224\n",
      "       225\n",
      "       226\n",
      "       227\n",
      "       228\n",
      "       229\n",
      "       230\n",
      "       231\n",
      "       232\n",
      "       233\n",
      "       234\n",
      "       235\n",
      "       236\n",
      "       237\n",
      "       238\n",
      "       239\n",
      "       240\n",
      "       241\n",
      "       242\n",
      "       243\n",
      "       244\n",
      "       245\n",
      "       246\n",
      "       247\n",
      "       248\n",
      "       249\n",
      "       250\n",
      "       251\n",
      "       252\n",
      "       253\n",
      "       253\n",
      "       hidden_gene alignment\n",
      "       [(0, 3941290, 50)] ACCCGCCCATGCCTCCCAAAGTGGTGGGATTATCGGTGTGACCCACCAAC\n",
      "       254\n",
      "       255\n",
      "       256\n",
      "       257\n",
      "       258\n",
      "       259\n",
      "       260\n",
      "       261\n",
      "       262\n",
      "       263\n",
      "       264\n",
      "       265\n",
      "       266\n",
      "       267\n",
      "       268\n",
      "       268\n",
      "       unaligned alignment\n",
      "       [] ATAGGGTGAGTCTCTGCAATGTTGGATGCAGCCTAACGTTTCAGACGACC\n",
      "       269\n",
      "       270\n",
      "       271\n",
      "       271\n",
      "       hidden_gene alignment\n",
      "       [(0, 6455465, 50)] GCGGGCGCCTGTAATCCCAGCTACAGCTGAGGCTGAGTCCGGAGGATCGC\n",
      "       272\n",
      "       273\n",
      "       274\n",
      "       275\n",
      "       276\n",
      "       276\n",
      "       hidden_gene alignment\n",
      "       [(0, 6455594, 50)] TCTCAAAAAAAAAAAGAAAAGAAAAAGAAAATTTGAGACGGAGTCTGGCT\n",
      "       277\n",
      "       278\n",
      "       279\n",
      "       280\n",
      "       281\n",
      "       281\n",
      "       unaligned alignment\n",
      "       [] TTCCCCCAGCGCCGGCGCCCTCTGTCCGTTATGTAGCCTGCTGGGAGATG\n",
      "       282\n",
      "       283\n",
      "       284\n",
      "       285\n",
      "       286\n",
      "       287\n",
      "       288\n",
      "       289\n",
      "       290\n",
      "       291\n",
      "       292\n",
      "       293\n",
      "       293\n",
      "       unaligned alignment\n",
      "       [] ATAAAAAGGAGTCCATTACCTACCGGCAGGACCAAGGCATCCTGTCTCGG\n",
      "       294\n",
      "       295\n",
      "       296\n",
      "       297\n",
      "       298\n",
      "       299\n",
      "       300\n",
      "       300\n",
      "       unaligned alignment\n",
      "       [] GCCCCGGAGAAGGTTACTAGATGACTCATCCGTGCTCTGGATCCGGTCGA\n",
      "       301\n",
      "       302\n",
      "       303\n",
      "       304\n",
      "       305\n",
      "       306\n",
      "       307\n",
      "       308\n",
      "       309\n",
      "       310\n",
      "       311\n",
      "       312\n",
      "       313\n",
      "       314\n",
      "       315\n",
      "       316\n",
      "       317\n",
      "       318\n",
      "       319\n",
      "       320\n",
      "       321\n",
      "       322\n",
      "       322\n",
      "       hidden_gene alignment\n",
      "       [(0, 2573320, 50)] GCAGAAGATTAATGTTTGGCAGAAGGAACTCGCAACAAGATTGTGTCCTT\n",
      "       323\n",
      "       323\n",
      "       hidden_gene alignment\n",
      "       [(0, 3941393, 50)] CAATGTAACTTAACAATTCAGTCTGAGTGAGTTTTGTAGTTTATAACATA\n",
      "       324\n",
      "       325\n",
      "       326\n",
      "       327\n",
      "       328\n",
      "       329\n",
      "       330\n",
      "       331\n",
      "       332\n",
      "       333\n",
      "       334\n",
      "       335\n",
      "       336\n",
      "       337\n",
      "       338\n",
      "       339\n",
      "       340\n",
      "       341\n",
      "       342\n",
      "       343\n",
      "       344\n",
      "       345\n",
      "       346\n",
      "       347\n",
      "       348\n",
      "       349\n",
      "       350\n",
      "       351\n",
      "       352\n",
      "       353\n",
      "       354\n",
      "       355\n",
      "       356\n",
      "       357\n",
      "       358\n",
      "       358\n",
      "       unaligned alignment\n",
      "       [] TATAGTCCATGTCGTTCTCTCTACGTGGAGATTCTAGGCGGTCGCAAAAG\n",
      "       359\n",
      "       360\n",
      "       361\n",
      "       362\n",
      "       363\n",
      "       364\n",
      "       365\n",
      "       366\n",
      "       367\n",
      "       368\n",
      "       369\n",
      "       370\n",
      "       371\n",
      "       372\n",
      "       373\n",
      "       374\n",
      "       375\n",
      "       376\n",
      "       377\n",
      "       378\n",
      "       379\n",
      "       380\n",
      "       381\n",
      "       382\n",
      "       383\n",
      "       384\n",
      "       385\n",
      "       386\n",
      "       387\n",
      "       388\n",
      "       389\n",
      "       390\n",
      "       390\n",
      "       hidden_gene alignment\n",
      "       [(0, 2574410, 50)] AAGAAACCAGGAGAAAAAGGAAGGTTTTGCGGTAGTCAGCTATGGTTTCC\n",
      "       391\n",
      "       392\n",
      "       392\n",
      "       hidden_gene alignment\n",
      "       [(0, 6514438, 50)] CACCTCCAAAGAGCCGTACGTCCACTGCCAGAAGCTACAACCGCAGTAAT\n",
      "       393\n",
      "       393\n",
      "       unaligned alignment\n",
      "       [] GTGACTTATGTAGGTCTATGGAGAAGTGCAGCTCACGGGAATAGCAAACC\n",
      "       394\n",
      "       395\n",
      "       396\n",
      "       397\n",
      "       398\n",
      "       399\n",
      "       400\n",
      "       401\n",
      "       402\n",
      "       403\n",
      "       404\n",
      "       405\n",
      "       406\n",
      "       407\n",
      "       408\n",
      "       409\n",
      "       410\n",
      "       411\n",
      "       412\n",
      "       413\n",
      "       413\n",
      "       unaligned alignment\n",
      "       [] GGGAGGGTTCAAATCGTCACATGAGCGTGCGAACTCCCCGGTCAGACTGT\n",
      "       414\n",
      "       415\n",
      "       415\n",
      "       hidden_gene alignment\n",
      "       [(0, 6461410, 50)] GCAAAAAGCCTCAGGTATTTCCCAAATCGGTAGCTGCAGAGGGGCTCTAT\n",
      "       416\n",
      "       417\n",
      "       418\n",
      "       419\n",
      "       420\n",
      "       421\n",
      "       422\n",
      "       423\n",
      "       424\n",
      "       425\n",
      "       426\n",
      "       427\n",
      "       428\n",
      "       429\n",
      "       430\n",
      "       430\n",
      "       hidden_gene alignment\n",
      "       [(0, 3941311, 50)] TGCTGGGATTATCGGTGCGAGCCACTAGCCCCTGCAAAGTCTACATCTCA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       431\n",
      "       432\n",
      "       433\n",
      "       434\n",
      "       435\n",
      "       436\n",
      "       437\n",
      "       438\n",
      "       438\n",
      "       unaligned alignment\n",
      "       [] TGAATCGCTTAATCCAACACTGAGGCCGGTCGTGGAGACGCGATGCAGTG\n",
      "       439\n",
      "       440\n",
      "       441\n",
      "       442\n",
      "       443\n",
      "       444\n",
      "       444\n",
      "       unaligned alignment\n",
      "       [] CCGAAGAGAAGCCACAAGTAGCGAGACCCGGAACTGGATAAGCTGCGATG\n",
      "       445\n",
      "       446\n",
      "       447\n",
      "       448\n",
      "       449\n",
      "       450\n",
      "       451\n",
      "       451\n",
      "       unaligned alignment\n",
      "       [] ACTTGTGTCCTTTGTCCCCTTGGAAGTTAGCTCGCGACGCTAGAATTTCA\n",
      "       452\n",
      "       453\n",
      "       454\n",
      "       454\n",
      "       unaligned alignment\n",
      "       [] GGTAAGGTATATCCTTCGCCGGTCACAACTCGAACAACGAGGATGTACGG\n",
      "       455\n",
      "       456\n",
      "       457\n",
      "       458\n",
      "       459\n",
      "       460\n",
      "       461\n",
      "       462\n",
      "       463\n",
      "       464\n",
      "       465\n",
      "       466\n",
      "       467\n",
      "       467\n",
      "       unaligned alignment\n",
      "       [] AGACCAATGCCTGGATACCATTCGCTACTAACAGTCCATAAAAATGTCGC\n",
      "       468\n",
      "       469\n",
      "       470\n",
      "       471\n",
      "       472\n",
      "       473\n",
      "       474\n",
      "       475\n",
      "       475\n",
      "       unaligned alignment\n",
      "       [] GTGTATTTTAAACTAAGTAAGCGGGAGTCCAGATGTAGTTGCACTTCCAT\n",
      "       476\n",
      "       477\n",
      "       478\n",
      "       479\n",
      "       480\n",
      "       481\n",
      "       482\n",
      "       483\n",
      "       484\n",
      "       484\n",
      "       unaligned alignment\n",
      "       [] GGTAGTGGCGTCACGAGGGGCACTATGGACCAAATGATTGGTAGAATGGG\n",
      "       485\n",
      "       486\n",
      "       487\n",
      "       488\n",
      "       489\n",
      "       490\n",
      "       491\n",
      "       492\n",
      "       493\n",
      "       494\n",
      "       495\n",
      "       496\n",
      "       497\n",
      "       498\n",
      "       499\n",
      "       499\n",
      "       hidden_gene alignment\n",
      "       [(0, 2564134, 50)] TTTGGCCTCTGCCAATGGAAATTCAGAAGTAGTACATCTCCTGCTGGACA\n",
      "       500\n",
      "       501\n",
      "       501\n",
      "       unaligned alignment\n",
      "       [] CGTGCTGGGTATAAAATAATGGAGCCCCGCCGTTTGATAAAACTTCGGGC\n",
      "       502\n",
      "       503\n",
      "       504\n",
      "       505\n",
      "       506\n",
      "       506\n",
      "       unaligned alignment\n",
      "       [] TTTTCAGCATAGTATACAATGCGTGAAATTACTTGTTGCTCCAATTCGAG\n",
      "       507\n",
      "       508\n",
      "       509\n",
      "       510\n",
      "       510\n",
      "       hidden_gene alignment\n",
      "       [(0, 6496086, 50)] AAACTGCCTATCTGCCATGCAAGTTTACGCTTAGTCCCGAAGATCTCGGA\n",
      "       511\n",
      "       512\n",
      "       513\n",
      "       514\n",
      "       515\n",
      "       516\n",
      "       517\n",
      "       518\n",
      "       519\n",
      "       520\n",
      "       521\n",
      "       522\n",
      "       523\n",
      "       524\n",
      "       525\n",
      "       526\n",
      "       527\n",
      "       528\n",
      "       528\n",
      "       hidden_gene alignment\n",
      "       [(0, 6455479, 50)] TCCCAGCTACAGGGGAGGCTGAGTCAGGAGAATCGCTTGAACCCTAGAGG\n",
      "       529\n",
      "       530\n",
      "       531\n",
      "       532\n",
      "       533\n",
      "       534\n",
      "       535\n",
      "       536\n",
      "       537\n",
      "       538\n",
      "       539\n",
      "       540\n",
      "       541\n",
      "       542\n",
      "       543\n",
      "       544\n",
      "       545\n",
      "       546\n",
      "       547\n",
      "       548\n",
      "       549\n",
      "       550\n",
      "       551\n",
      "       552\n",
      "       553\n",
      "       554\n",
      "       555\n",
      "       556\n",
      "       557\n",
      "       558\n",
      "       559\n",
      "       560\n",
      "       561\n",
      "       562\n",
      "       563\n",
      "       564\n",
      "       565\n",
      "       566\n",
      "       567\n",
      "       568\n",
      "       568\n",
      "       hidden_gene alignment\n",
      "       [(0, 2580032, 50)] ATGAATTCCACAAAGGAGAAGCAGAAAACCTGAAAAACGGCAATATCCTG\n",
      "       569\n",
      "       570\n",
      "       570\n",
      "       unaligned alignment\n",
      "       [] TGTGTGGATTATTACCGAGGACAAGGAGTGACTGTTCGTTTGGCACTTCA\n",
      "       571\n",
      "       571\n",
      "       unaligned alignment\n",
      "       [] GGAGCGCTTTGATGCCTTTGGTGATGAGCTGGAGGCATCACACATAGTGC\n",
      "       572\n",
      "       573\n",
      "       574\n",
      "       575\n",
      "       576\n",
      "       576\n",
      "       unaligned alignment\n",
      "       [] CGACACCCAGTATTAGAGCGCGTTAACTCCCCCCAAATGGAAAGGCAAGC\n",
      "       577\n",
      "       578\n",
      "       579\n",
      "       580\n",
      "       581\n",
      "       582\n",
      "       583\n",
      "       584\n",
      "       585\n",
      "       586\n",
      "       587\n",
      "       588\n",
      "       589\n",
      "       590\n",
      "       591\n",
      "       592\n",
      "       593\n",
      "       594\n",
      "       595\n",
      "       596\n",
      "       597\n",
      "       598\n",
      "       599\n",
      "       600\n",
      "       601\n",
      "       602\n",
      "       603\n",
      "       604\n",
      "       605\n",
      "       606\n",
      "       607\n",
      "       608\n",
      "       609\n",
      "       609\n",
      "       unaligned alignment\n",
      "       [] CTACTGAGGAAGCAAGAATTGAAGTTGCTGAAACTAGCTTATTCTCACCT\n",
      "       610\n",
      "       611\n",
      "       612\n",
      "       613\n",
      "       614\n",
      "       615\n",
      "       616\n",
      "       617\n",
      "       618\n",
      "       618\n",
      "       unaligned alignment\n",
      "       [] AAACGTTAGGCTGCTCCCTAGGTCTGAACACTTCTCCGTCATTCCGCCCC\n",
      "       619\n",
      "       620\n",
      "       621\n",
      "       622\n",
      "       623\n",
      "       624\n",
      "       625\n",
      "       625\n",
      "       hidden_gene alignment\n",
      "       [(0, 3941385, 50)] CACTTATACAATGTAACTGACCAATTCAGTCTGAGTGAGTTTTGTAGTTT\n",
      "       626\n",
      "       627\n",
      "       628\n",
      "       629\n",
      "       630\n",
      "       631\n",
      "       632\n",
      "       633\n",
      "       634\n",
      "       635\n",
      "       636\n",
      "       637\n",
      "       638\n",
      "       639\n",
      "       640\n",
      "       641\n",
      "       641\n",
      "       unaligned alignment\n",
      "       [] GGGCGATATACGAACGTTCTATCAGACATAGGGAATGAAGGCCAAGAACG\n",
      "       642\n",
      "       643\n",
      "       644\n",
      "       645\n",
      "       646\n",
      "       647\n",
      "       648\n",
      "       648\n",
      "       unaligned alignment\n",
      "       [] TCATTAGTTCTCAACGGTGCCTGGCGTATCCTATGTCACTCCACAATAGC\n",
      "       649\n",
      "       650\n",
      "       650\n",
      "       unaligned alignment\n",
      "       [] GGTTAGAGTACTACCAATTTCCGGAGGTCTGTAAGGGCGAAGACGTATAC\n",
      "       651\n",
      "       652\n",
      "       653\n",
      "       654\n",
      "       655\n",
      "       656\n",
      "       657\n",
      "       657\n",
      "       unaligned alignment\n",
      "       [] AACAAAAGCAGATAGATGAGGCTGAACAGAAAATTAATTCAGAGTTTCTC\n",
      "       658\n",
      "       659\n",
      "       659\n",
      "       unaligned alignment\n",
      "       [] CACATCCGCGTCAGTAGGGTACGAATTTGTCCCTGAGCTGGTCGCGGGTA\n",
      "       660\n",
      "       661\n",
      "       662\n",
      "       663\n",
      "       664\n",
      "       665\n",
      "       666\n",
      "       667\n",
      "       668\n",
      "       669\n",
      "       670\n",
      "       671\n",
      "       672\n",
      "       673\n",
      "       674\n",
      "       675\n",
      "       676\n",
      "       677\n",
      "       678\n",
      "       679\n",
      "       680\n",
      "       681\n",
      "       682\n",
      "       683\n",
      "       684\n",
      "       685\n",
      "       686\n",
      "       687\n",
      "       688\n",
      "       689\n",
      "       690\n",
      "       691\n",
      "       692\n",
      "       693\n",
      "       694\n",
      "       695\n",
      "       696\n",
      "       697\n",
      "       698\n",
      "       699\n",
      "       700\n",
      "       701\n",
      "       702\n",
      "       703\n",
      "       704\n",
      "       705\n",
      "       706\n",
      "       707\n",
      "       707\n",
      "       hidden_gene alignment\n",
      "       [(0, 6455505, 50)] GGAGACTCGCTTGAACCGTGGTGGTGAAGCTTGCAGTGAGCCGAGATCGC\n",
      "       708\n",
      "       709\n",
      "       710\n",
      "       710\n",
      "       unaligned alignment\n",
      "       [] TCGTGTGTTGGGGGGCTCATCATTCCGCTGGATATGGTGCACGTTAAAAG\n",
      "       711\n",
      "       712\n",
      "       713\n",
      "       714\n",
      "       715\n",
      "       716\n",
      "       717\n",
      "       718\n",
      "       719\n",
      "       719\n",
      "       unaligned alignment\n",
      "       [] AAATCCTAAAGCTGTCCAAATCTGAACGACAAAGCTTCGGCCAATTGGCC\n",
      "       720\n",
      "       721\n",
      "       722\n",
      "       723\n",
      "       724\n",
      "       725\n",
      "       726\n",
      "       727\n",
      "       728\n",
      "       729\n",
      "       729\n",
      "       unaligned alignment\n",
      "       [] TGGATGAACCTGGTAATATTTATAAATATTGACTCCAGATTTTAGGACAG\n",
      "       730\n",
      "       731\n",
      "       732\n",
      "       733\n",
      "       734\n",
      "       735\n",
      "       736\n",
      "       737\n",
      "       738\n",
      "       739\n",
      "       740\n",
      "       741\n",
      "       742\n",
      "       743\n",
      "       744\n",
      "       745\n",
      "       745\n",
      "       hidden_gene alignment\n",
      "       [(0, 2558955, 50)] GACTGGGCTGGGTGTTTCCTTCGGAGGGGGGGCTTGGCCTTTCCTGCGGT\n",
      "       746\n",
      "       747\n",
      "       748\n",
      "       749\n",
      "       750\n",
      "       751\n",
      "       752\n",
      "       753\n",
      "       753\n",
      "       unaligned alignment\n",
      "       [] CATGGCGCGCCGACCAGTATGGGTTCGCATCGAAAAACGATATGTGGTAG\n",
      "       754\n",
      "       755\n",
      "       756\n",
      "       757\n",
      "       758\n",
      "       759\n",
      "       760\n",
      "       760\n",
      "       unaligned alignment\n",
      "       [] CATCCGTTCAGACGCTTCTCGTCCTCTCTCATAATCCTGAGATATTGCTT\n",
      "       761\n",
      "       762\n",
      "       763\n",
      "       764\n",
      "       765\n",
      "       766\n",
      "       767\n",
      "       768\n",
      "       769\n",
      "       770\n",
      "       771\n",
      "       772\n",
      "       773\n",
      "       774\n",
      "       775\n",
      "       776\n",
      "       777\n",
      "       778\n",
      "       779\n",
      "       779\n",
      "       hidden_gene alignment\n",
      "       [(0, 3941376, 50)] CAAGGGTATCACTTACACAATGTAACTTACCAATTCAGTCTGAGTGAGTT\n",
      "       780\n",
      "       781\n",
      "       782\n",
      "       783\n",
      "       784\n",
      "       785\n",
      "       786\n",
      "       786\n",
      "       hidden_gene alignment\n",
      "       [(0, 2590402, 50)] AAGATATTGCCATGCTAAGACTGGAACTAGATGAAACAATACATAAGTAC\n",
      "       787\n",
      "       788\n",
      "       789\n",
      "       790\n",
      "       791\n",
      "       792\n",
      "       793\n",
      "       794\n",
      "       795\n",
      "       796\n",
      "       797\n",
      "       798\n",
      "       799\n",
      "       800\n",
      "       801\n",
      "       802\n",
      "       803\n",
      "       804\n",
      "       805\n",
      "       806\n",
      "       807\n",
      "       808\n",
      "       809\n",
      "       810\n",
      "       811\n",
      "       812\n",
      "       813\n",
      "       814\n",
      "       815\n",
      "       816\n",
      "       817\n",
      "       818\n",
      "       819\n",
      "       820\n",
      "       821\n",
      "       822\n",
      "       823\n",
      "       824\n",
      "       825\n",
      "       826\n",
      "       827\n",
      "       827\n",
      "       unaligned alignment\n",
      "       [] CCCCCACCGAAGAGCGGCAGAATAGGACTCAGACTGTAAGCTTGTAGTTT\n",
      "       828\n",
      "       829\n",
      "       830\n",
      "       831\n",
      "       832\n",
      "       833\n",
      "       834\n",
      "       835\n",
      "       836\n",
      "       836\n",
      "       hidden_gene alignment\n",
      "       [(0, 3941405, 50)] CCACTTCAGTCTGACTGAGTTTTGTAGTTTATAACATATTGATGAAAACA\n",
      "       837\n",
      "       838\n",
      "       839\n",
      "       840\n",
      "       841\n",
      "       842\n",
      "       843\n",
      "       844\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       845\n",
      "       846\n",
      "       847\n",
      "       848\n",
      "       849\n",
      "       850\n",
      "       851\n",
      "       852\n",
      "       853\n",
      "       854\n",
      "       855\n",
      "       856\n",
      "       856\n",
      "       hidden_gene alignment\n",
      "       [(0, 6455659, 50)] GAGTGCAGTGGCGCGATCCCGGCCCACTGCAAGCTCTGCCTCCCGGGTTC\n",
      "       857\n",
      "       858\n",
      "       859\n",
      "       860\n",
      "       861\n",
      "       862\n",
      "       863\n",
      "       864\n",
      "       865\n",
      "       866\n",
      "       867\n",
      "       868\n",
      "       869\n",
      "       870\n",
      "       871\n",
      "       872\n",
      "       873\n",
      "       874\n",
      "       875\n",
      "       876\n",
      "       877\n",
      "       878\n",
      "       879\n",
      "       880\n",
      "       881\n",
      "       882\n",
      "       883\n",
      "       884\n",
      "       885\n",
      "       886\n",
      "       887\n",
      "       888\n",
      "       889\n",
      "       890\n",
      "       891\n",
      "       892\n",
      "       893\n",
      "       894\n",
      "       895\n",
      "       896\n",
      "       897\n",
      "       898\n",
      "       898\n",
      "       unaligned alignment\n",
      "       [] TTTATGAGTGATCTTTGTATGGTTAAGGACGACTCACAGTCTCCGAAGCC\n",
      "       899\n",
      "       900\n",
      "       901\n",
      "       902\n",
      "       903\n",
      "       904\n",
      "       905\n",
      "       905\n",
      "       unaligned alignment\n",
      "       [] CAACCTCATATGGATAGACCGGAGGGAAAGCTGTCCGAGTGGAACTGCGT\n",
      "       906\n",
      "       907\n",
      "       907\n",
      "       unaligned alignment\n",
      "       [] TTGGCCGCCTCAATTAAGATCGCCACGGCGCTGCGGGCCCTTACAACCCT\n",
      "       908\n",
      "       909\n",
      "       910\n",
      "       911\n",
      "       912\n",
      "       913\n",
      "       914\n",
      "       915\n",
      "       916\n",
      "       917\n",
      "       918\n",
      "       919\n",
      "       920\n",
      "       920\n",
      "       unaligned alignment\n",
      "       [] ACTAGCCAAGTGGCCGCACGCAGCATTGAAAGATCTACTTGGCGCTACGA\n",
      "       921\n",
      "       922\n",
      "       923\n",
      "       924\n",
      "       925\n",
      "       926\n",
      "       927\n",
      "       927\n",
      "       hidden_gene alignment\n",
      "       [(0, 6455617, 50)] AAAGAAAATTTGAGACGGAGTCTGGCTCTGTCGCCCAGGCTGGAGTGCAG\n",
      "       928\n",
      "       929\n",
      "       930\n",
      "       931\n",
      "       932\n",
      "       933\n",
      "       934\n",
      "       935\n",
      "       936\n",
      "       936\n",
      "       unaligned alignment\n",
      "       [] CACTTCGCAGAGATCCGTACCTTCCAACCCATCTACGTATTCACCTGTGT\n",
      "       937\n",
      "       937\n",
      "       unaligned alignment\n",
      "       [] CTGCCAGGGTTCTCACTCACTAGGGGCGGACTAGCCGCATCTGCTTAACT\n",
      "       938\n",
      "       939\n",
      "       940\n",
      "       941\n",
      "       942\n",
      "       943\n",
      "       944\n",
      "       945\n",
      "       946\n",
      "       947\n",
      "       948\n",
      "       949\n",
      "       950\n",
      "       951\n",
      "       952\n",
      "       953\n",
      "       954\n",
      "       955\n",
      "       956\n",
      "       956\n",
      "       unaligned alignment\n",
      "       [] ATCGAATTTGCTGCTGTATTATGAAGGTCCCGTTTATGCACTACCCTGGA\n",
      "       957\n",
      "       958\n"
     ]
    }
   ],
   "source": [
    "total_time = 0\n",
    "for i in range(len(reads)):\n",
    "    read_sequence = reads[i][1]\n",
    "    t = -time()\n",
    "    alignment = aligner.align(read_sequence)\n",
    "    total_time += time() + t\n",
    "    print('      ', i)\n",
    "    if alignment == -1:\n",
    "        unable_to_match.append(i)\n",
    "        print('      ', str(i) + ' not matched!')\n",
    "    else:\n",
    "        case, _ = evaluate_alignment(genome_sequence, read_sequence, alignment, unknown_isoforms, genome_isoform_offsets)\n",
    "        if case != CASE_GENE:\n",
    "            print('      ', i)\n",
    "            print('      ', case, 'alignment')\n",
    "            print('      ', alignment, read_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
